### PR TITLE
Editorial: clarify string representation of numbers with radix ≠ 10

### LINKED ISSUE #2847: How “implementation-defined” is %Number.prototype.toString% meant to be?
The final step given for [Number.prototype.toString](https://tc39.es/ecma262/#sec-number.prototype.tostring) says:

> 6. Return the String representation of this [Number value](https://tc39.es/ecma262/#number-value) using the radix specified by radixMV. Letters a-z are used for digits with values 10 through 35. The precise algorithm is [implementation-defined](https://tc39.es/ecma262/#implementation-defined), however the algorithm should be a generalization of that specified in [6.1.6.1.20](https://tc39.es/ecma262/#sec-numeric-types-number-tostring).

The “generalization of...” bit is only qualified with “should” (and I’m not certain how much “generalization of” conveys anyway). It surprised me to learn that this has been left pretty open-ended mainly because I wasn’t aware of engine differences that might have motivated it. BigInt inherited the same pattern.

Could/should the exact behavior be specified? If not, what prevented it?

### PR DESCRIPTION
**Pull request by michaelficarra**
Fixes #2847 by adding a `radix` parameter to `Number::toString`.

/cc @bathos

### CODE REVIEW

File: spec.html
Comment (bakkot) [inline spec.html:2200]: You should update the NOTE which refers to this step also, since it still assumes base 10.

**Review by bakkot — APPROVED** [review]
LGTM other than comment but I want to wait for @syg's approval as well.

File: spec.html
Comment (michaelficarra) [reply spec.html:2200]: Done.

**Comment by bathos** [comment]
@michaelficarra [%Bigint.prototype.toString%](https://tc39.es/ecma262/#sec-bigint.prototype.tostring) uses the same “should be a generalization” text and points at [BigInt::toString](https://tc39.es/ecma262/#sec-numeric-types-bigint-tostring). As far as I can tell the algorithm-in-practice aligns with Number-when-not-base-10. Would make sense (and would it be desirable) for them to point to a common algorithm that’s agnostic to the specific numeric language type?

(This PR already stands on its own with plenty of value, so I’m not suggesting it should address anything related to %BigInt.prototype.toString% — just curious about how that could end up looking.)

File: spec.html
Comment (gibson042) [inline spec.html:2219]: This alias could be more clearly named.
```suggestion
            1. If _n_ &lt; 0, then
              1. Let _exponentSign_ be the code unit 0x002D (HYPHEN-MINUS).
            1. Else,
              1. Let _exponentSign_ be the code unit 0x002B (PLUS SIGN).
```

**Review by gibson042 — APPROVED** [review]
I like the improved generality and precision, and this does a decent job of containing the decimal-specific scale weirdness. That said, I do wonder if the latter could be achieved even better, e.g.
```
1. If _radix_ &ne; 10 or _n_ is in the inclusive interval from -5 to 21, then
  1. If _n_ &ge; _k_, return the string-concatenation of…
  1. If _n_ &gt; 0, return the string-concatenation of…
  1. Else,
    1. Assert: _n_ &le; 0.
    1. Return the string-concatenation of…
1. Assert: _radix_ is 10.
1. …
```

File: spec.html
Comment (michaelficarra) [reply spec.html:2201]: I don't know if we ever actually decided that, but we are at least consistent at the moment (aside from ℝ, ℤ, 𝔽). 

See https://github.com/tc39/ecmarkup/issues/173#issuecomment-658447485

**Comment by michaelficarra** [comment]
@gibson042 Thanks for the suggestions. I took all of them. Please re-review to check my work.

@bathos Agree, we could do something similar about `%Bigint.prototype.toString%`. Can you open an issue to track it? I don't think they should share an algorithm, but we can at least make a similar clarification.

File: spec.html
Comment (jmdyck) [inline spec.html:2199]:
```suggestion
            1. [id="step-number-tostring-intermediate-values"] Otherwise, let _n_, _k_, and _s_ be integers such that _k_ &ge; 1, _radix_<sup>_k_ - 1</sup> &le; _s_ &lt; _radix_<sup>_k_</sup>, 𝔽(_s_ &times; _radix_<sup>_n_ - _k_</sup>) is _x_, and _k_ is as small as possible. Note that _k_ is the number of digits in the representation of _s_ using radix _radix_, that _s_ is not divisible by _radix_, and that the least significant digit of _s_ is not necessarily uniquely determined by these criteria.
```

File: spec.html
Comment (jmdyck) [inline spec.html:31272]:
```suggestion
          1. Return Number::toString(_x_, _radixMV_).
```

**Comment by michaelficarra** [comment]
Thanks @jmdyck. Done.

File: spec.html
Comment (gibson042) [inline spec.html:2199]: Unnecessary "otherwise".
```suggestion
            1. [id="step-number-tostring-intermediate-values"] Let _n_, _k_, and _s_ be integers such that _k_ &ge; 1, _radix_<sup>_k_ - 1</sup> &le; _s_ &lt; _radix_<sup>_k_</sup>, 𝔽(_s_ &times; _radix_<sup>_n_ - _k_</sup>) is _x_, and _k_ is as small as possible. Note that _k_ is the number of digits in the representation of _s_ using radix _radix_, that _s_ is not divisible by _radix_, and that the least significant digit of _s_ is not necessarily uniquely determined by these criteria.
```

**Review by gibson042 — APPROVED** [review]

**Review by syg — APPROVED** [review]

File: spec.html
Comment (bakkot) [inline spec.html:2217]:
```suggestion
            1. NOTE: In this case, the input will be represented using `1.2e+3`-style scientific E notation.
            1. Assert: _radix_ is 10.
```

Comment (michaelficarra) [reply spec.html:2217]: Slightly tweaked the wording to not imply that this is the one and only format.

Comment (bakkot) [reply spec.html:2217]: Works for me.

File: spec.html
Comment (michaelficarra) [inline spec.html:2192]: FYI this spelling of "zeros" is different from what we had before but I'm unable to determine what the OED says, if anything, about the preferred spelling. I will change it back to "zeroes" for now.

**Pull request merged**
